### PR TITLE
feat: credentials framework

### PR DESCRIPTION
**Pull request by g-linville**
You can read more about it in the doc I am adding in this PR, but basically this adds a "credentials framework" so that we have an easy way to supply secret values in environment variables to tools that need them. First you need to have a credential provider tool that prints a JSON object in the following format: `{"env":{"ENV_VAR_1": "value1", "ENV_VAR_2": "value2"}}`. Then, you can reference the credential tool in a script like this:

```
name: my-tool
credential: my-credential-tool.gpt

(tool body here)
```

The key-value pairs of environment variables that `my-credential-tool.gpt` returns will be added to the environment for the main tool's execution.

### CODE REVIEW

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -207,7 +212,7 @@ func (e *Engine) Start(ctx Context, input string) (*Return, error) {
 		InternalSystemPrompt: tool.Parameters.InternalPrompt,
 	}
 
-	for _, subToolName := range tool.Parameters.Tools {
+	for _, subToolName := range append(tool.Parameters.Tools, tool.Parameters.Credentials...) {
```
Comment (ibuildthecloud) [inline pkg/engine/engine.go:215]: Why do we give the credential tools to the LLM?

File: pkg/credentials/credential.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package credentials
+
+import (
+	"encoding/json"
+
+	"github.com/docker/cli/cli/config/types"
+)
+
+type Credential struct {
+	ToolID string            `json:"toolID"`
```
Comment (ibuildthecloud) [inline pkg/credentials/credential.go:10]: Should this be toolname or tool? Not id.  FYI, ID's are not really "portable" They are basically generated on demand and not really guaranteed to be stable over time.  Only make sense in the context of a single "Program"

File: pkg/engine/engine.go (hunk 1 above)
Comment (g-linville) [reply pkg/engine/engine.go]: Fixed

File: pkg/credentials/credential.go (hunk 2 above)
Comment (g-linville) [reply pkg/credentials/credential.go]: Fixed and renamed to `ToolName`. Though based on a discussion that I had with Taylor earlier today, we are probably going to need to use something else as the identifier/unique name for each credential

File: pkg/credentials/credential.go [hunk 3]
```diff
@@ -31,12 +33,26 @@ func credentialFromDockerAuthConfig(authCfg types.AuthConfig) (Credential, error
...
+		Context:  ctx,
+		ToolName: tool,
 		Env:      env,
 	}, nil
 }
 
-func (c Credential) String() string {
-	return c.ToolName
+func toolNameWithCtx(toolName, credCtx string) string {
+	return toolName + "///" + credCtx
```
Comment (g-linville) [inline pkg/credentials/credential.go:49]: Open for suggestions for something better than this. I need a way of combining toolName and credCtx such that it still creates a valid URL.

File: pkg/runner/runner.go [hunk 4]
```diff
@@ -278,3 +298,67 @@ func recordStateMessage(state *engine.State) error {
...
+		return nil, fmt.Errorf("failed to create credentials store: %w", err)
+	}
+
+	for _, credToolName := range callCtx.Tool.Credentials {
+		cred, exists, err := store.Get(credToolName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get credentials for tool %s: %w", credToolName, err)
+		}
+
+		// If the credential doesn't already exist in the store, run the credential tool in order to get the value,
```
Comment (iwilltry42) [inline pkg/runner/runner.go:319]: Not sure if that'd be too much, but maybe we can have a flag for "reprompt" to update all cached credentials in a run? My laziness coming through.. :D

File: pkg/cli/gptscript.go [hunk 5]
```diff
@@ -38,28 +38,29 @@ type GPTScript struct {
...
+	SubTool           string `usage:"Use tool of this name, not the first tool in file" local:"true"`
+	Assemble          bool   `usage:"Assemble tool to a single artifact, saved to --output" hidden:"true" local:"true"`
+	ListModels        bool   `usage:"List the models available and exit" local:"true"`
+	ListTools         bool   `usage:"List built-in tools and exit" local:"true"`
+	Server            bool   `usage:"Start server" local:"true"`
+	ListenAddress     string `usage:"Server listen address" default:"127.0.0.1:9090" local:"true"`
+	Chdir             string `usage:"Change current working directory" short:"C"`
+	Daemon            bool   `usage:"Run tool as a daemon" local:"true" hidden:"true"`
+	Ports             string `usage:"The port range to use for ephemeral daemon ports (ex: 11000-12000)" hidden:"true"`
+	CredentialContext string `usage:"Context name in which to store credentials" default:"default" short:"c"`
```
Comment (ibuildthecloud) [inline pkg/cli/gptscript.go:56]: Don't use `-c`.  Just make it long and verbose right now.  `-c` might be useful for something else more important later

File: pkg/cli/credential.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/spf13/cobra"
+)
+
+type Credential struct {
+	root        *GPTScript
+	AllContexts bool `short:"A" usage:"List credentials for all contexts" local:"true"`
+}
+
+func NewCredential(root *GPTScript) *cobra.Command {
+	c := cmd.Command(&Credential{root: root}, cobra.Command{
```
Comment (ibuildthecloud) [inline pkg/cli/credential.go:21]: I've been using the Customize interface approach to change the default cobra.Command fields.  I feel like that is slightly less confusing pattern. I'm going to probably drop support for this style of passing it as a second arg.

https://github.com/gptscript-ai/gptscript/blob/main/pkg/cli/gptscript.go#L102

File: pkg/cli/credential.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	"github.com/acorn-io/cmd"
+	"github.com/gptscript-ai/gptscript/pkg/config"
+	"github.com/gptscript-ai/gptscript/pkg/credentials"
+	"github.com/spf13/cobra"
+)
+
+type Credential struct {
+	root        *GPTScript
+	AllContexts bool `short:"A" usage:"List credentials for all contexts" local:"true"`
```
Comment (ibuildthecloud) [inline pkg/cli/credential.go:17]: Don't do `-A`.  Either have no short or if you really want a short value then `-a`.  The `-A` is very kubernetes-y.

File: pkg/runner/runner.go (hunk 4 above)
Comment (g-linville) [reply pkg/runner/runner.go:319]: I asked Darren about it and he wants to hold off on this for now.

File: pkg/cli/credential_delete.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+type Delete struct {
+	root *GPTScript
+}
+
+func NewDelete(root *GPTScript) *cobra.Command {
+	return cmd.Command(&Delete{root: root}, cobra.Command{
+		Use:          "delete <tool name>",
+		SilenceUsage: true,
+		Short:        "Delete a stored credential",
+		Args:         cobra.ExactArgs(1),
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/cli/credential_delete.go:21]: For now, no.

File: pkg/cli/gptscript.go (hunk 5 above)
Comment (g-linville) [reply pkg/cli/gptscript.go]: Fixed

File: pkg/cli/credential.go (hunk 6 above)
Comment (g-linville) [reply pkg/cli/credential.go]: Fixed

File: pkg/cli/credential.go (hunk 7 above)
Comment (g-linville) [reply pkg/cli/credential.go]: Fixed

File: docs/docs/03-tools/04-credentials.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+Here is a simple example of a credential provider tool that uses the builtin `sys.prompt` to ask the user for some input:
+
+```yaml
+# my-credential-tool.gpt
+name: my-credential-tool
+
+#!/usr/bin/env bash
+
+output=$(gptscript -q --cache=false sys.prompt '{"message":"Please enter your fake credential.","fields":"credential"}')
```
[the remark this answers is not in the document]
Comment (g-linville) [reply docs/docs/03-tools/04-credentials.md]: Fixed

**Review by iwilltry42 — APPROVED** [review]

**Comment by g-linville** [comment]
(I got verbal approval from Darren to merge.)

**Pull request merged**